README.md

```diff
@@ -1,486 +1,679 @@
+  }
 }
 ```
 
-is roughly equivalent to
+The class definition would be desugared roughly as follows:
 
 ```js
-class C {
-  method() { }
+let initialize;
+class MyElement extends HTMLElement {
+  clickHandler() { }
+  constructor(...args) {
+    super(...args);
+    initialize.call(this);
+  }
 }
-f(C, "method");
+{method: MyElement.prototype.clickHandler, initialize} =
+  on('click')(MyElement.prototype.clickHandler,
+              {kind: "init-method", isStatic: false, name: "clickHandler"});
 ```
 
-Details:
-- `@register` can be used on any method, field, accessor, or the class as a whole.
-- Arguments passed into the callback given to `@register`:
-  - First argument: the "target": For static fields and methods, or the class itself, it is the class; for instance fields and methods, it is the class's prototype.
-  - Second argument: For public fields, methods or accessors, the property key; for private, or for the class itself, only one argument is passed.
-  - Note, there is no third argument; the property descriptor is not passed into the callback, but the callback could look it up itself.
-- The return value of the callback must be undefined.
-- All `@wrap` callbacks run before all `@register` callbacks. This is because `@wrap` is used to set up the class, and `@register` runs on the class after it's created.
-
-### `@initialize`
+#### `@bound` with `init`
 
-The `@initialize` decorator intercepts the initialization of a public class field and runs a callback supplied to the decorator in place of `Object.defineProperty`. For example:
+The `init` keyword for methods can also be used to build a `@bound` decorator, used as follows:
 
 ```js
 class C {
-  @initialize(f) a = b;
+  #x = 1;
+  @bound init method() { return this.#x; }
 }
+
+let c = new C;
+let m = c.method;
+m();  // 1, not TypeError
 ```
 
-is roughly equivalent to the following:
+The `@bound` decorator can be implemented as follows:
 
 ```js
-class C {
-  constructor() {
-    f.call(this, f, b, "a");
-  }
+function bound(method, {kind, name}) {
+  assert(kind === "init-method");
+  return {method, initialize() { this[name] = this[name].bind(this); }};
 }
 ```
 
-Details:
-- `@initialize` can only be used on public field declarations (including symbols/computed property names), not private.
-- `@initialize` can be used with both static and instance fields.
-- The return value of the callback must be undefined.
+## Possible extensions
+
+Decorators on further constructs are investigated in [EXTENSIONS.md](./EXTENSIONS.md).
 
-## User-defined Decorators
+# Syntax
 
-JavaScript programmers can make their own decorators by composing built-in decorators.
+This decorators proposal uses the syntax of the previous Stage 2 decorators proposal. This means that:
+- Decorator expressions are restricted to a chain of variables, property access with `.` but not `[]`, and calls `()`. To use an arbitrary expression as a decorator, `@(expression)` is an escape hatch.
```

Comment:
if this escape hatch exists, why is it desired/valuable to require it?

Reply:
Sorry, I don't understand the question. Could you rephrase it?

Reply:
If it is possible to use any expression in a decorator, what is the reason why some forms require wrapping parens?

Reply:
To avoid ambiguity with computed property keys. (We had some extensive bikeshedding here in the stage 2 decorator days...)